sw/device/tests/otbn/otbn_rsa_test.c

```diff
@@ -0,0 +1,903 @@
+      kDifOtbnOk) {
+    return kOtbnRsaError;
+  }
+
+  if (dif_otbn_dmem_write(&ctx->dif, ctx->otbnptr_modulus, modulus,
+                          size_bytes) != kDifOtbnOk) {
+    return kOtbnRsaError;
+  }
+
+  if (dif_otbn_dmem_write(&ctx->dif, ctx->otbnptr_exp, private_exponent,
+                          size_bytes) != kDifOtbnOk) {
+    return kOtbnRsaError;
+  }
+
+  if (dif_otbn_dmem_write(&ctx->dif, ctx->otbnptr_in, in, size_bytes) !=
+      kDifOtbnOk) {
+    return kOtbnRsaError;
+  }
+
+  // Call OTBN to perform operation
+  otbn_rsa_result_t rv;
+  rv = otbn_call_function(ctx, &kOtbnFuncRsaRsaDecrypt);
+  if (rv != kOtbnRsaOk) {
+    return rv;
+  }
+
+  rv = otbn_wait_for_done(ctx);
+  if (rv != kOtbnRsaOk) {
+    return rv;
+  }
+
+  // Read back result
+  if (dif_otbn_dmem_read(&ctx->dif, ctx->otbnptr_out, out, size_bytes) !=
+      kDifOtbnOk) {
+    return kOtbnRsaError;
+  }
+
+  return kOtbnRsaOk;
+}
+
+static void check_data(const uint8_t *actual, const uint8_t *expected,
+                       size_t size_bytes) {
+  for (int i = 0; i < size_bytes; ++i) {
+    CHECK(actual[i] == expected[i],
+          "Data mismatch at byte %d: 0x%x (actual) != 0x%x (expected)", i,
+          actual[i], expected[i]);
+  }
+}
+
+uint64_t t_start;
+uint64_t t_end;
+
+static void profile_start() { t_start = ibex_mcycle_read(); }
+
+static void profile_end(const char *msg) {
+  t_end = ibex_mcycle_read();
+  uint32_t cycles = t_end - t_start;
+  uint32_t time_us = cycles / 100;
+  LOG_INFO("%s took %u cycles or %u us @ 100 MHz.", msg, cycles, time_us);
+}
+
+static void rsa_roundtrip(uint32_t size_bytes, const uint8_t *modulus,
+                          const uint8_t *private_exponent, const uint8_t *in,
+                          const uint8_t *encrypted_expected,
+                          uint8_t *out_encrypted, uint8_t *out_decrypted) {
+  dif_otbn_config_t otbn_config = {
+      .base_addr = mmio_region_from_addr(TOP_EARLGREY_OTBN_BASE_ADDR),
+  };
+
+  otbn_rsa_t otbn_rsa_ctx;
+
+  // Initialize
+  profile_start();
+  CHECK(otbn_rsa_init(&otbn_rsa_ctx, &otbn_config) == kOtbnRsaOk);
+  CHECK(otbn_rsa_load_app(&otbn_rsa_ctx) == kOtbnRsaOk);
+  profile_end("Initialization");
+
+  // Encrypt
+  LOG_INFO("Encrypting");
+  profile_start();
+  CHECK(otbn_rsa_encrypt(&otbn_rsa_ctx, modulus, in, out_encrypted,
+                         size_bytes) == kOtbnRsaOk);
+  check_data(out_encrypted, encrypted_expected, size_bytes);
+  profile_end("Encryption");
+
+  if (kTestDecrypt) {
+    // Decrypt
+    LOG_INFO("Decrypting");
+    profile_start();
+    CHECK(otbn_rsa_decrypt(&otbn_rsa_ctx, modulus, private_exponent,
+                           encrypted_expected, out_decrypted,
+                           size_bytes) == kOtbnRsaOk);
+    check_data(out_decrypted, in, size_bytes);
+    profile_end("Decryption");
+  }
+}
+
+static void test_rsa512_roundtrip() {
+  static const uint8_t modulus[kRsa512SizeBytes] = {
```

Comment:
It's a static const, so it should be kModulus. Applies to others in this function and beyond.